Document persistent context and file storage (#1465)
flowforge/flowforge: docs/user/filenodes.md

# FlowForge File Nodes

Projects running within FlowForge include a set of nodes that make it possible

> in a containered environment to read and write to persistent storage.

Reviewer: ```suggestion
to store files safely regardless of the environment. For edge devices the files are stored locally, containerized applications will read and write to persistent storage.
```

Author: outdated

The File Nodes are a clone of the original Node-RED nodes so that flows written
for the cloud can be ran the same on a device which uses its own local filesystem.

For that reason, these nodes cannot be used at the same time as the built-in 
Node-RED file nodes. To use these nodes, you must exclude 10-file-js in the 

> `nodesExcludes` array in your settings file.

Reviewer: I don't have access to these settings? How does an end user exclude anything?

Author: outdated

### Nodes

There are two nodes in the File Node collection:

- `file` - A file node for writing to persistent storage
- `file in` - A file node for reading from persistent storage

The nodes send the whole `msg` object between projects. Due to the way the nodes
encode messages, there are some data types that do not get sent. For example,
the `msg.req`/`msg.res` properties used by the core HTTP nodes will not be sent.

> ### Requirements

Reviewer: How does an end user know if these requirements are met?

Author: outdated

Also, the re-write is more inline with other premium features where "requirements" are not totally explicit due to being implemented in FlowForge Cloud

> A persistent storage backend provided by a @flowforge/file-server instance

Reviewer: ```suggestion
```

Author: outdated

> ### GitHub

Reviewer: Remove this section, a user doesn't care IMO.

Author: outdated

Also, for the file nodes, this info is left in same as the other CE nodes docs

# FlowForge File Nodes

Projects running within FlowForge include a set of nodes that make it possible
to store files safely regardless of the environment. For edge devices the files 
are stored locally, containerized applications can read and write to persistent 
storage via API calls to a file-server instance by using these nodes.

> *NOTE: This set of nodes allow you to read and write persistent files in a containerised

Reviewer: User documentation; how do I as a user know this?

Reviewer: @hardillb Please consider end-users each time you're implementing config options or multiple backends. It keeps creating unclear situations and it's not great for anyone except the super-power users. We should improve the lives of the 95% of non-power users first after we're able to improve the last 5% too.

Persistent context has now multiple choices, files too, going forward let's have one and only one implementation.

environment only when coupled with a [file-storage](../install/file-storage/README.md) instance*

## Nodes

There are two nodes in the File Node collection:

- `file` - A file node for writing to persistent storage
- `file in` - A file node for reading from persistent storage

## Install

### FlowForge

> The File Nodes are installed automatically with FlowForge however in order for a 

Reviewer: ```suggestion
The File Nodes are installed automatically with FlowForge. However in order for a 
```

Author: done (next commit)

project to use them, the built-in Node-RED file nodes must first be disabled. 

> This is done by simply adding `10-file-js` in the **Exclude nodes by filename** 

Reviewer: This is not true, I cannot edit this.

<img width="463" alt="Screenshot 2022-12-20 at 09 19 54" src="https://user-images.githubusercontent.com/2529595/208617817-99ef70d8-4cc7-464d-9a4e-76731ddbab76.png">

Author: It is true however your particular account is not permitted to edit this.

How should this be worded?

#### it could be simplified to...
" The File Nodes are installed automatically with FlowForge however they must be enabled by a project owner so long as the platform administrator has permitted this in the project template"

####  or it could be more technical like...

" The File Nodes are installed automatically with FlowForge however in order for a 
project to use them, the built-in Node-RED file nodes must first be disabled. 
This is done by simply adding `10-file-js` in the **Exclude nodes by filename** 
section of your FlowForge projects settings under  the **Palette** section.
**NOTE:**  The **Exclude nodes by filename** setting can only be modified by a project owner and the project template must permit an owner to edit this field. The project template can only be edited by an administrator of the platform. "

Please advise.

Author: revised in new commits

section of your FlowForge projects settings under  the **Palette** section

> ### Standalone Node-RED

Reviewer: Why would a user go this route? I don't get this section at all.

Author: As the FlowForge instance may not be on FF Cloud, the FF admin / owner needs to know how to do this.

But you do raise an interesting point. There may be cases where they want to disable the node-red file nodes but NOT enable the cloud based file nodes.

Need to discuss with @hardillb - but this may be a next iteration thing.

Reviewer 2: Current situation:

- File storage can be enabled/disabled globally for a FlowForge Instance by adding/removing the `fileStore` block in the `flowforge.yml`
- If enabled then disabling the core file nodes will enable the FlowForge File Nodes
- There is currently no way to have projects with a mix of some with node file nodes at all and some with the FlowForge Nodes.

I suggest we raise an item for the next release to implment disabling all File node support in the template if it's enabled at the FlowForge Insance leve.

Author: Story raised: https://github.com/flowforge/flowforge/issues/1472

These nodes can be installed through the Manage Palette option in the Node-RED
editor, or on the command-line:

```bash
cd ~/.node-red
npm install @flowforge/nr-file-nodes
```

> This assumes the default location of the Node-RED user directory. If you are not

Reviewer: If users just use the command palette this isn't a question right? If so; why recommend this?

Author: Actually, I have just checked & `npm install` is currently the only (recommended) way to install the file nodes - as it is not published in the flows library.

For now, I will remove the part about using the node-red library

@hardillb @knolleary should we publish `@flowforge/nr-file-nodes` on the flows lib?

Author: Done (next commit)

Author: Following discussion with Ben, these will not be published to flows & install information will be removed in next commit.
